Added Mac OS support

BadrGhanem opened the pull request:
I have implemented support for Mac OS and addressed several compilation errors that were preventing the code from compiling correctly. Comprehensive testing has been performed to ensure the functionality and reliability of the changes made.

BadrGhanem committed:
Added Mac OS support

I have implemented support for Mac OS and addressed several compilation errors that were preventing the code from compiling correctly. Comprehensive testing has been performed to ensure the functionality and reliability of the changes made.

javalikescript commented on webview-cocoa.c:
```diff
@@ -122,64 +133,103 @@ static void run_save_panel(id self, SEL cmd, id download, id filename,
 }
 
+
+static const char *parse_HTML_content(const char *html, int *comma_index) {
```
The name is misleading as there is no HTML parsing, maybe `parse_data_URI_content_type` would be more precise

BadrGhanem replied:
I agree 'parse_data_URI_content_type' is better name to use

javalikescript commented on webview-cocoa.c:
```diff
@@ -45,73 +49,80 @@
         if (result == (id)NSModalResponseOK) {
-          completionHandler(objc_msgSend(openPanel, sel_registerName("URLs")));
+
+                  id urls = ((id(*)(id, SEL))objc_msgSend)(openPanel, sel_registerName("URLs"));
```
Please fix indentation

BadrGhanem replied:
no problem, will do

javalikescript commented on webview-cocoa.c:
```diff
@@ -335,92 +391,110 @@ WEBVIEW_API int webview_init(struct webview *w) {
+     
+
+
+   int comma_index;
```
Please fix indentation

BadrGhanem replied:
no problem, will do

javalikescript commented on webview-cocoa.c:
```diff
@@ -335,92 +391,110 @@ WEBVIEW_API int webview_init(struct webview *w) {
+  const char *MIMEType = parse_HTML_content(w->url, &comma_index);
+
+  if (MIMEType != NULL)
+  {
```
I would prefer to stick to the file curly brackets style which seems to be on the same line

BadrGhanem replied:
no problem

javalikescript commented on webview-cocoa.c:
```diff
@@ -122,64 +133,103 @@ static void run_save_panel(id self, SEL cmd, id download, id filename,
+
+    int i = 0;
+    while (*p != ',' && *p != '\0' && i < 255) {
+        if (!isspace(*p)) {
```
Is there a reason to ignore spaces and to accept missing comma?
I would prefer to be as much as possible restrictive in parsing and to not change the URI content.
I propose something like the following:
```c
static const char *parse_data_URI(const char *uri, int *comma_index) {
  if (uri == NULL || comma_index == NULL || strncmp(uri, "data:", 5) != 0) {
    return NULL;
  }
  const char *pstart = uri + 5;
  const char *pend = uri + 255;
  const char *p = pstart;
  while (*p != ',') {
    if (*p <= ' ' || *p >= '\x7f' || p > pend) {
      return NULL;
    }
    p++;
  }
  int size = (int)(p - pstart);
  char *result = malloc(size + 1);
  if (result != NULL) {
    strncpy(result, pstart, size);
    result[size] = '\0';
    *comma_index = (int)(p - uri);
  }
  return result;
}
```

BadrGhanem replied:
The intention is to ensure its robustness against incorrect usage such as:

"   data :text/html, Some content here"
"   da ta:text/html, Some content here"
"   d a ta :text/html, Some content here"

"data:text/ html, Some content here"
"data :text/html, Some content here"
"data: text/ html, Some content here"

Otherwise, it will not return 'text/html'. 
I can accept the proposed solution - however, the proposed solution might still encounter failures as per the examples provided so have to mention that in the readme.

javalikescript replied:
Both solution are good for me, the difference is rather subjective, so I am accepting yours

javalikescript reviewed (commented):
Thank you for your work !
I am not able to compile nor run on Mac OS 😞 but I have some comments

BadrGhanem commented:
You are welcome! I was positively drawn to this C library.
I am waiting for your confirmation on my responses so that I can push an updated version that we both agree upon 😄

BadrGhanem committed:
Update webview-cocoa.c

Addressing code review comments. This commit -
The previous 'parse_data_URI' logic has not been changed yet. However, I will proceed with the changes after receiving confirmation from the author.

BadrGhanem commented:
I have addressed your code review comments by making a new commit to the pull request. Kindly take a look at it when you have a moment.

BadrGhanem commented:
> Thank you for your work ! I am not able to compile nor run on Mac OS 😞 but I have some comments

Btw:  Here is a screenshot of the compiled process successfully running on Mac OS 😄
![macos_support](https://github.com/javalikescript/webview-c/assets/17663219/1be6c177-2fc2-4229-a985-d3ff9232888a)

BadrGhanem commented:
It seems that I accidentally requested a pull request review again, sorry about that!
